### PR TITLE
Upgrade contracts to v0.3.0 with better tests

### PR DESCRIPTION
**Pull request by alpe**
Upgrade contracts to v0.3.0
Full IBC enabled end to end test of stake, unstake

### CODE REVIEW

File: tests/e2e/mvp_test.go [hunk 1]
```diff
@@ -36,195 +38,171 @@ func TestMVP(t *testing.T) {
...
-	require.NoError(t, err)
+
+	execMsg = fmt.Sprintf(`{"stake_remote":{"contract":"%s", "amount": {"denom":%q, "amount":"%d"}, "msg":%q}}`,
+		providerContracts.externalStaking.String(),
+		sdk.DefaultBondDenom, 40_000_000,
+		base64.StdEncoding.EncodeToString([]byte(fmt.Sprintf(`{"validator": "%s"}`, myExtValidatorAddr))))
+	providerCli.MustExecVault(execMsg)
+
+	require.NoError(t, coord.RelayAndAckPendingPackets(ibcPath))
+	assert.NotEqual(t, 50_000_000, providerCli.QueryVaultFreeBalance()) // todo: set correct number
```
Comment (alpe) [inline tests/e2e/mvp_test.go:126]: This fails. The free amount should be reduced

**Comment by ethanfrey** [comment]
Are you really using alpha.1 contracts, not beta.2? It would be good to update the PR name.

It would also be good to have a download script there that also writes `version.txt` file to the testdata directory

File: tests/e2e/test_client.go [hunk 2]
```diff
@@ -93,20 +93,23 @@ func (p TestProviderClient) mustExec(contract sdk.AccAddress, payload string, fu
...
 	resp, err := p.chain.SendMsgs(&wasmtypes.MsgExecuteContract{
 		Sender:   p.chain.SenderAccount.GetAddress().String(),
 		Contract: contract.String(),
 		Msg:      []byte(payload),
 		Funds:    funds,
 	})
 	require.Error(p.t, err, "Response: %v", resp)
+	// workaround sequence issue until wasmd testchain code is updated
+	p.chain.NextBlock()
+	require.NoError(p.t, p.chain.SenderAccount.SetSequence(p.chain.SenderAccount.GetSequence()+1))
```
Comment (ethanfrey) [inline tests/e2e/test_client.go:112]: Ah, very cool

File: tests/e2e/mvp_test.go [hunk 3]
```diff
@@ -25,206 +27,194 @@ func TestMVP(t *testing.T) {
...
+	// update system time
+	coord.CurrentTime = releasedAt.Add(time.Minute)
+	coord.UpdateTime()
+	coord.CommitBlock(providerChain, consumerChain)
+
+	providerCli.MustExecExtStaking(`{"withdraw_unbonded":{}}`)
+	assert.Equal(t, 50_000_000, providerCli.QueryVaultFreeBalance())
+
+	// provider chain
+	// ==============
```
Comment (pinosu) [inline tests/e2e/mvp_test.go:211]: 👍

File: tests/e2e/test_client.go [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+func NewProviderClient(t *testing.T, chain *ibctesting.TestChain) *TestProviderClient {
+	return &TestProviderClient{t: t, chain: chain}
+}
+
+type ProviderContracts struct {
+	vault           sdk.AccAddress
+	externalStaking sdk.AccAddress
+}
+
+func (p *TestProviderClient) BootstrapContracts(connId, portID string) ProviderContracts {
```
Comment (pinosu) [inline tests/e2e/test_client.go:49]: very nice 👍

**Review by pinosu — APPROVED** [review]
LGTM 👍

**Pull request merged**